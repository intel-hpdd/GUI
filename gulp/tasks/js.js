//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2016 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.

'use strict';

var paths = require('../paths.json');
var gulp = require('gulp');
var babel = require('gulp-babel');
var sourcemaps = require('gulp-sourcemaps');
var destDir = require('../dest-dir');
var relativeSourcemapsSource = require('gulp-relative-sourcemaps-source');

function jsDeps (fn) {
  return gulp.src(paths.js.deps, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.jsDepsDev = function jsDepsDev () {
  return jsDeps(jsDepsDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsDepsProd = function jsDepsProd () {
  return jsDeps(jsDepsProd)
  .pipe(gulp.dest('./dest'));
};

function socketWorker (fn) {
  return gulp.src(paths.js.socketWorker, {
    since: gulp.lastRun(fn)
  })
  .pipe(sourcemaps.init({ loadMaps: true }))
  .pipe(sourcemaps.write('.'));
}

exports.socketWorkerDev = function socketWorkerDev () {
  return socketWorker(socketWorkerDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.socketWorkerProd = function socketWorkerProd () {
  return socketWorker(socketWorkerProd)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.dest('./dist'));
};

function systemConfig (fn) {
  return gulp.src(paths.js.config, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.systemConfigDev = function systemConfigDev () {
  return systemConfig(systemConfigDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

var babelDev = babel.bind(null, {
  presets: [],
  babelrc: false,
  plugins: [
    'check-es2015-constants',
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    'transform-async-to-generator',
    'transform-es2015-classes',
    'transform-es2015-modules-systemjs'
  ]
});

var babelProd = babel.bind(null, {
  presets: ['es2015'],
  babelrc: false,
  plugins: [
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    [
      'angularjs-annotate',
      {
        explicitOnly: true
      }
    ],
    'transform-es2015-modules-systemjs'
  ]
});

function jsSource (fn) {
  return gulp.src(paths.js.source, {
    since: gulp.lastRun(fn),
    base: '.'
  })
  .pipe(sourcemaps.init());
}

exports.jsSourceDev = function jsSourceDev () {
  return jsSource(jsSourceDev)
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '/static/chroma_ui' }))
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsSourceProd = function jsSourceProd () {
  return jsSource(jsSourceProd)
  .pipe(babelProd())
  .pipe(relativeSourcemapsSource({dest: 'dest'}))
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTest = function jsTest () {
  return gulp.src(paths.js.tests, {
    since: gulp.lastRun(jsTest),
    base: '.'
  })
  .pipe(sourcemaps.init())
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTestFixtures = function jsTestFixtures () {
  return gulp.src(paths.js.testFixtures, {
    since: gulp.lastRun(jsTestFixtures),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};

exports.jsTestDeps = function jsTestDeps () {
  return gulp.src(paths.js.testDeps, {
    since: gulp.lastRun(jsTestDeps),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};
