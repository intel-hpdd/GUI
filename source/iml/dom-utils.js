// @flow

type QuerySelector<T: HTMLElement> = (HTMLElement | Document, string) => T;

export const querySelector: QuerySelector<*> = (
  elm: HTMLElement | Document,
  sel: string
) => {
  const selected = elm.querySelector(sel);

  if (!selected)
    throw new Error(
      `Could not find selector ${sel} for element ${elm.nodeName}`
    );

  return selected;
};
