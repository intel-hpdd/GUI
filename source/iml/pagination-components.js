// @flow

//
// Copyright (c) 2018 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

import type { Meta } from "./api-types.js";
import Inferno, { linkEvent } from "inferno";
import WindowClickListener from "./window-click-listener.js";
import DropdownContainer from "./dropdown-component.js";

type EntriesProps = {
  entries: number,
  setEntries: Function
};

export const EntriesDropdown = (props: EntriesProps) => (
  <WindowClickListener>
    <DropdownContainer>
      <button type="button" class="btn btn-info btn-sm dropdown-toggle" aria-haspopup="true" aria-expanded="false">
        Entries: {props.entries} <span class="caret" />
      </button>
      <ul role="menu" class="dropdown-menu">
        {[10, 25, 50, 100].map(x => (
          <li>
            <a onClick={linkEvent(x, props.setEntries)}>{x}</a>
          </li>
        ))}
      </ul>
    </DropdownContainer>
  </WindowClickListener>
);

type PerPageProps = {
  meta: Meta
};

export const TableInfo = ({ meta }: PerPageProps) => (
  <span>
    showing {meta.offset + 1}-{Math.min(meta.offset + meta.limit, meta.total_count)} of {meta.total_count}
  </span>
);

const computePage = ({ limit, offset }) => (limit === 0 ? 1 : Math.floor(offset / limit + 1));

const computePages = ({ limit, total_count: totalCount }) => (limit === 0 ? 1 : Math.ceil(totalCount / limit));

const MAX_SIZE = 5;

function getPages(page: number, pages: number): number[] {
  let startPage = 1;
  let endPage = pages;

  if (pages > MAX_SIZE) {
    startPage = Math.max(page - Math.floor(MAX_SIZE / 2), 1);
    endPage = startPage + MAX_SIZE - 1;

    if (endPage > pages) {
      endPage = pages;
      startPage = endPage - MAX_SIZE + 1;
    }
  }

  const xs = [];
  for (let i = startPage; i <= endPage; i++) xs.push(i);
  return xs;
}

const computeOffset = (page, limit) => (page - 1) * limit;

type PagerProps = {
  meta: Meta,
  setOffset: Function
};

export const Pager = ({ meta, setOffset }: PagerProps) => {
  const pages = computePages(meta);

  if (pages === 1) return;

  const page = computePage(meta);

  return (
    <ul class="pagination">
      <li className={`pagination-prev ${page === 1 ? "disabled" : ""}`}>
        <a onClick={linkEvent(computeOffset(page - 1, meta.limit), setOffset)}>‹</a>
      </li>
      {getPages(page, pages).map(x => (
        <li className={`pagination-page ${page === x ? "active" : ""}`}>
          <a onClick={linkEvent(computeOffset(x, meta.limit), setOffset)}>{x}</a>
        </li>
      ))}
      <li class={`pagination-next ${page === pages ? "disabled" : ""}`}>
        <a onClick={linkEvent(computeOffset(page + 1, meta.limit), setOffset)}>›</a>
      </li>
    </ul>
  );
};
