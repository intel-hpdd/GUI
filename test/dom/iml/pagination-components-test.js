// @flow
import { render } from "inferno";
import { Pager, EntriesDropdown, TableInfo } from "../../../source/iml/pagination-components.js";
import { renderToSnapshot } from "../../test-utils.js";
import { querySelector } from "../../../source/iml/dom-utils.js";
import type { Meta } from "../../../source/iml/api-types.js";

describe("EntriesDropdown DOM testing", () => {
  let clickHandler, vnode;

  beforeEach(() => {
    clickHandler = jest.fn();
    vnode = <EntriesDropdown entries={10} setEntries={clickHandler} />;
  });

  it("should render entries as expected", () => {
    expect(renderToSnapshot(vnode)).toMatchSnapshot();
  });

  describe("clicking", () => {
    let root;

    beforeEach(() => {
      root = document.createElement("div");
      querySelector(document, "body").appendChild(root);
      render(vnode, root);
    });

    afterEach(() => {
      querySelector(document, "body").removeChild(root);
    });

    it("should open when clicked", () => {
      querySelector(root, "button").click();

      expect(root.innerHTML).toMatchSnapshot();
    });

    it("should call handler when entry is clicked", () => {
      querySelector(root, "button").click();
      root.querySelectorAll("a")[1].click();

      expect(clickHandler).toHaveBeenCalledOnceWith(25, expect.any(Object));
    });
  });
});

describe("TableInfo DOM testing", () => {
  let meta: Meta;

  beforeEach(() => {
    meta = {
      limit: 10,
      total_count: 100,
      offset: 10,
      previous: null,
      next: null
    };
  });

  it("should render as expected", () => {
    expect(renderToSnapshot(<TableInfo meta={meta} />)).toMatchSnapshot();
  });
});

describe("Pager DOM testing", () => {
  let meta: Meta, clickHandler;

  beforeEach(() => {
    meta = {
      limit: 10,
      total_count: 100,
      offset: 10,
      previous: null,
      next: null
    };

    clickHandler = jest.fn();
  });

  it("should render nothing if there is a single page", () => {
    meta = {
      limit: 10,
      total_count: 10,
      offset: 0,
      previous: null,
      next: null
    };
    expect(renderToSnapshot(<Pager meta={meta} setOffset={clickHandler} />)).toMatchSnapshot();
  });

  it("should render as expected", () => {
    const newMeta = {
      ...meta,
      limit: 25,
      total_count: 100,
      offset: 49
    };

    expect(renderToSnapshot(<Pager meta={newMeta} setOffset={clickHandler} />)).toMatchSnapshot();
  });

  it("should reset pages if on the trailing edge", () => {
    const newMeta = {
      ...meta,
      limit: 25,
      total_count: 200,
      offset: 199
    };

    expect(renderToSnapshot(<Pager meta={newMeta} setOffset={clickHandler} />)).toMatchSnapshot();
  });

  it("should disable the previous button if on the first page", () => {
    const newMeta = {
      ...meta,
      limit: 25,
      total_count: 100,
      offset: 0
    };

    expect(renderToSnapshot(<Pager meta={newMeta} setOffset={clickHandler} />)).toMatchSnapshot();
  });

  it("should disable the next button if on the last page", () => {
    const newMeta = {
      ...meta,
      limit: 25,
      total_count: 100,
      offset: 95
    };

    expect(renderToSnapshot(<Pager meta={newMeta} setOffset={clickHandler} />)).toMatchSnapshot();
  });

  describe("click handling", () => {
    let root;

    beforeEach(() => {
      root = document.createElement("div");
      querySelector(document, "body").appendChild(root);

      render(<Pager meta={meta} setOffset={clickHandler} />, root);
    });

    afterEach(() => {
      querySelector(document, "body").removeChild(root);
    });

    it("should go to the previous page when clicking previous", () => {
      querySelector(root, ".pagination-prev a").click();

      expect(clickHandler).toHaveBeenCalledOnceWith(0, expect.any(Object));
    });

    it("should go to the next page when clicking next", () => {
      querySelector(root, ".pagination-next a").click();

      expect(clickHandler).toHaveBeenCalledOnceWith(20, expect.any(Object));
    });

    it("should go to the clicked page", () => {
      querySelector(root, ".pagination-page a").click();

      expect(clickHandler).toHaveBeenCalledOnceWith(0, expect.any(Object));
    });
  });
});
