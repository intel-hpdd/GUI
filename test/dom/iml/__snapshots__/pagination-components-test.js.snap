// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntriesDropdown DOM testing clicking should open when clicked 1`] = `"<div class=\\"btn-group dropdown open\\"><button class=\\"btn btn-info btn-sm dropdown-toggle\\" type=\\"button\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\">Entries: 10 <span class=\\"caret\\"></span></button><ul class=\\"dropdown-menu\\" role=\\"menu\\"><li><a>10</a></li><li><a>25</a></li><li><a>50</a></li><li><a>100</a></li></ul></div>"`;

exports[`EntriesDropdown DOM testing should render entries as expected 1`] = `
<div>
  <div
    class="btn-group dropdown "
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn btn-info btn-sm dropdown-toggle"
      type="button"
    >
      Entries: 
      10
       
      <span
        class="caret"
      />
    </button>
    <ul
      class="dropdown-menu"
      role="menu"
    >
      <li>
        <a>
          10
        </a>
      </li>
      <li>
        <a>
          25
        </a>
      </li>
      <li>
        <a>
          50
        </a>
      </li>
      <li>
        <a>
          100
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Pager DOM testing should disable the next button if on the last page 1`] = `
<div>
  <ul
    class="pagination"
  >
    <li
      class="pagination-prev "
    >
      <a>
        ‹
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        1
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        2
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        3
      </a>
    </li>
    <li
      class="pagination-page active"
    >
      <a>
        4
      </a>
    </li>
    <li
      class="pagination-next disabled"
    >
      <a>
        ›
      </a>
    </li>
  </ul>
</div>
`;

exports[`Pager DOM testing should disable the previous button if on the first page 1`] = `
<div>
  <ul
    class="pagination"
  >
    <li
      class="pagination-prev disabled"
    >
      <a>
        ‹
      </a>
    </li>
    <li
      class="pagination-page active"
    >
      <a>
        1
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        2
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        3
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        4
      </a>
    </li>
    <li
      class="pagination-next "
    >
      <a>
        ›
      </a>
    </li>
  </ul>
</div>
`;

exports[`Pager DOM testing should render as expected 1`] = `
<div>
  <ul
    class="pagination"
  >
    <li
      class="pagination-prev "
    >
      <a>
        ‹
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        1
      </a>
    </li>
    <li
      class="pagination-page active"
    >
      <a>
        2
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        3
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        4
      </a>
    </li>
    <li
      class="pagination-next "
    >
      <a>
        ›
      </a>
    </li>
  </ul>
</div>
`;

exports[`Pager DOM testing should render nothing if there is a single page 1`] = `
<div>
  
</div>
`;

exports[`Pager DOM testing should reset pages if on the trailing edge 1`] = `
<div>
  <ul
    class="pagination"
  >
    <li
      class="pagination-prev "
    >
      <a>
        ‹
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        4
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        5
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        6
      </a>
    </li>
    <li
      class="pagination-page "
    >
      <a>
        7
      </a>
    </li>
    <li
      class="pagination-page active"
    >
      <a>
        8
      </a>
    </li>
    <li
      class="pagination-next disabled"
    >
      <a>
        ›
      </a>
    </li>
  </ul>
</div>
`;

exports[`TableInfo DOM testing should render as expected 1`] = `
<div>
  <span>
    showing 
    11
    -
    20
     of 
    100
  </span>
</div>
`;
